### DISCUSSION TITLE
Re-run failed uploads?

### DISCUSSION THREAD

**Discussion by caphector** [aaronwmorris/indi-allsky#1379] [Q&A]
I've noticed that sometimes the daily files will fail to transfer (star trails, keograms, timelapses). Is there a way to re-run these uploads or should I use something like rsync?

**Comment by aaronwmorris** [comment]
The status of file transfers, S3 uploads, and SyncAPI transfers are all tracked in the database.  The `./misc/upload_sync.py` will reattempt any upload that failed.

```
source virtualenv/indi-allsky/bin/activate

./misc/upload_sync.py
```

**Reply by aaronwmorris** [reply]
@caphector Did this work for you?

**Reply by caphector** [reply]
I just had a chance to try; have been a bit busy. Seems to be getting consistent timeouts:

```2024-07-11 12:29:35,691 [INFO] MainProcess generic.connect() [52]: Connecting to example.com (22) as caphector with pycurl_sftp
2024-07-11 12:29:35,692 [INFO] MainProcess generic.put() [66]: Uploading /mnt/stars/allsky/ccd_54528347-52c5-49e9-9fe2-65c3a007035f/timelapse/20240704/allsky-timelapse_ccd1_20240704_day.mp4
2024-07-11 12:29:35,693 [INFO] MainProcess pycurl_sftp.put() [124]: pycurl URL: sftp://example.com:22/public_html/allsky/videos/allsky-timelapse_ccd1_20240704_day.mp4
2024-07-11 12:29:38,418 [INFO] MainProcess upload_sync.addUploadEntries() [240]: Adding 11 upload entries (8266 remaining)
2024-07-11 12:29:45,694 [ERROR] MainProcess uploader.processUpload() [456]: Connection failure: Operation timed out after 10000 milliseconds with 0 bytes received```

Test my basic CLI SFTP and I was able to connect normally; each one seemed to fail in this way.

Tried giving only --upload-images and it did videos as well; didn't see a way to only do one or the other but I might have failed to read the docs.

**Reply by aaronwmorris** [reply]
The 10s timeout is the connect timeout.  This means the remote side is not accepting the connection at Layer 4 (or you have a bad network connection).  It sounds like the remote side may have firewall rules that limit a single connection from each source IP.  Even with a single upload, it may take 1-2 seconds after each upload for the remote side to recognize your connection is no longer there.

I may need to introduce the ability to sleep between each transfer.

**Reply by aaronwmorris** [reply]
Actually, before I do this.  Is indi-allsky running (and uploading) at the same time as trying to run the upload sync?

**Reply by caphector** [reply]
It was running when I tried to do the sync.

**Reply by aaronwmorris** [reply]
If indi-allsky is performing uploads, and your hosting provider is limiting connections, you may have to stop indi-allsky while you perform a sync.

**Reply by caphector** [reply]
I don't *think* I have limited connections; I upload to a VPS and I was able to make four separate SFTP connections via `sftp` on my CLI. I also tested the connection by making a direct `sftp` connection to my host while Indi-Allsky was running.

**Comment by gaitskell** [comment]
I have found that many .FITS files didn't successfuly copy from the local raspi, up to GCP (Google Cloud Platform). The much smaller .jpg's are present.  I believe my WiFi link to the raspi may have been substandard for a few nights. I am going to try upload_sync.py suggestion to see if the FITS missing from GCP get filled in. (I will do this during daytime when no live images are uploading.)

**Reply by gaitskell** [reply]
Advice for others:
I am maintaining a GCP [Google Cloud] archive of all the FITS and JPG from nightly sessions, plus the diagnostic images/movies for night's viewing.

I need to fill in the files that are missing in GCP. 

I used the following CLI commands. This also provides a log of transfers in .log file

```
cd ~/indi-allsky/
source virtualenv/indi-allsky/bin/activate
./misc/upload_sync.py sync
```
Note argument **sync**

Initially I had 
`webUI > Config > Object Storage > Connect Timeout
`
set to 10 secs. This created errors in the transfers for mp4.
So I increased this to 60 secs in the WebUI and the transfers could complete within that time.
